wpilibj/src/main/java/edu/wpi/first/wpilibj/shuffleboard/Shuffleboard.java

```diff
@@ -0,0 +1,268 @@
+  static NetworkTableInstance instance = NetworkTableInstance.getDefault();
+
+  private static List<BuilderBase> m_builders = new ArrayList<>();
+  private static List<SendableData> m_sendables = new ArrayList<>();
+
+  static {
+    HLUsageReporting.reportSmartDashboard(); // TODO use a unique ID for shuffleboard?
+  }
+
+  private static final class SendableData {
+
+    private final SendableBuilder m_sendable;
+    private final SendableBuilderImpl m_builder;
+    private boolean m_initialized = false;
+
+    SendableData(SendableBuilder sendable, SendableBuilderImpl builder) {
+      this.m_sendable = sendable;
+      this.m_builder = builder;
+    }
+
+    /**
+     * Initializes the sendable, if not already initialized. No-op if the sendable has been
+     * initialized.
+     */
+    public void initialize() {
+      if (!m_initialized) {
+        m_builder.setTable(instance.getTable(m_sendable.generateKey()));
+        m_sendable.getSendable().initSendable(m_builder);
+        m_builder.startListeners();
+        m_initialized = true;
+      }
+    }
+
+    /**
+     * Updates the sendable.
+     */
+    public void update() {
+      m_builder.updateTable();
+    }
+  }
+
+  /**
+   * Gets the base Shuffleboard table in the current NetworkTables instance. This is a method
+   * instead of stored as a field because the instance can change during tests, and a field would
+   * still reference the original instance.
+   */
+  private static NetworkTable getBaseTable() {
+    return instance.getTable(BASE_TABLE_NAME);
+  }
+
+  /**
+   * Updates all the values in Shuffleboard. This should be called repeatedly in a loop (for
+   * example, in an iterative or timed robot's autonomousPeriodic and teleopPeriodic methods).
+   */
+  public static void update() {
+    NetworkTable baseTable = getBaseTable();
+    // Update tabs
+    NetworkTable tabsTable = baseTable.getSubTable(".tabs");
+    m_builders.stream()
+        .map(b -> b.getTab().getName())
+        .sorted()
+        .forEach(tabName -> {
+          NetworkTable subTable = tabsTable.getSubTable(tabName);
+          subTable.getEntry("Name").setString(tabName);
+          subTable.getEntry("AutopopulatePrefix").setString("/Shuffleboard/" + tabName + "/");
+        });
+
+    // Update layout and widget metadata
+    NetworkTable metadata = baseTable.getSubTable(".metadata");
+    for (BuilderBase builder : m_builders) {
+      NetworkTable tabTable = metadata.getSubTable(builder.getTab().getName());
+
+      // Add layout metadata (type, title, properties)
+      NetworkTable parent = updateLayoutMetadata(builder, tabTable);
+
+      // Add widget metadata (type, properties)
+      updateWidgetMetadata(builder, parent);
+    }
+
+    // Update sendables
+    // Basic entries do not need to be updated by us; they are set by users via the
+    // NetworkTableEntry instances they get when completing an add() operation
+    m_sendables.forEach(data -> {
```

Comment:
Does this allocate a lambda per call? Why not just a standard loop? Since this is called once per robot loop, we should avoid allocations as much as possible.

Reply:
It's a non-capturing lambda, so there's only one object allocated.